crates/polars-core/src/chunked_array/logical/datetime.rs

```diff
@@ -31,63 +31,66 @@ impl LogicalType for DatetimeChunked {
+                    Some(m) => Ok((self.0.as_ref() * m)
+                        .into_datetime(*to_unit, tz.clone())
+                        .into_series()),
+                    // scale to lower precision (eg: ns → us, ns → ms, us → ms)
+                    None => match divisor {
+                        Some(d) => Ok(self
+                            .0
+                            .apply_values(|v| v.div_euclid(d))
+                            .into_datetime(*to_unit, tz.clone())
+                            .into_series()),
+                        None => unreachable!("must always have a time unit divisor here"),
+                    },
+                };
+                result
             },
             #[cfg(feature = "dtype-date")]
-            (Datetime(tu, _), Date) => match tu {
-                TimeUnit::Nanoseconds => Ok((self.0.as_ref() / NS_IN_DAY)
-                    .cast(&Int32)
-                    .unwrap()
-                    .into_date()
-                    .into_series()),
-                TimeUnit::Microseconds => Ok((self.0.as_ref() / US_IN_DAY)
-                    .cast(&Int32)
-                    .unwrap()
-                    .into_date()
-                    .into_series()),
-                TimeUnit::Milliseconds => Ok((self.0.as_ref() / MS_IN_DAY)
-                    .cast(&Int32)
-                    .unwrap()
-                    .into_date()
-                    .into_series()),
+            (Datetime(tu, _), Date) => {
+                let cast_to_date = |tu_in_day: i64| {
+                    let mut dt = self
+                        .0
+                        .apply_values(|v| v.div_euclid(tu_in_day))
+                        .cast(&Int32)
+                        .unwrap()
+                        .into_date()
+                        .into_series();
+                    dt.set_sorted_flag(self.is_sorted_flag());
+                    Ok(dt)
+                };
+                match tu {
+                    TimeUnit::Nanoseconds => cast_to_date(NS_IN_DAY),
+                    TimeUnit::Microseconds => cast_to_date(US_IN_DAY),
+                    TimeUnit::Milliseconds => cast_to_date(MS_IN_DAY),
+                }
             },
             #[cfg(feature = "dtype-time")]
             (Datetime(tu, _), Time) => Ok({
-                let (modder, multiplier) = match tu {
+                let (scaled_mod, multiplier) = match tu {
                     TimeUnit::Nanoseconds => (NS_IN_DAY, 1i64),
                     TimeUnit::Microseconds => (US_IN_DAY, 1_000i64),
                     TimeUnit::Milliseconds => (MS_IN_DAY, 1_000_000i64),
                 };
                 self.0
-                    .apply_values(|v| (v % modder * multiplier) + (NS_IN_DAY * (v < 0) as i64))
+                    .apply_values(|v| {
+                        let t = v % scaled_mod * multiplier;
+                        t + (NS_IN_DAY * (v < 0 && t != 0) as i64)
```

Comment:
```suggestion
                        t + (NS_IN_DAY * (t < 0) as i64)
```

Reply:
That panics for me in the "test_cast_to_time_and_combine" test with `ValueError('hour must be in 0..23')` 🤔